### PR TITLE
fix: correct URL when switching page modes

### LINKED ISSUE #4615: In DesignPage set URL by reading the designPageLocationState or to the root bot's main dialog
###  Steps to recreate

In the context of a botproject with a root bot and skill, click on a dialog inside a skill the URL is set correctly `(bot/{​​​​​​​rootBotId}​​​​​​​/skill/{​​​​​​​skillId}​`​​​​​​). Navigate using the sidebar to a different page and come back to the design page. The url is set at `bot/{​​​​​​​skillId}​​​​​​​/dialogs/*`. This URL is wrong as a skill needs to be nested inside the RootBotId in the url always.

### Solution

  
Read the designPageLocationState and set the URL from that state or at a minimum navigate to the RootBot's main dialog whenever the user selects the DesignPage icon on the left nav

### PR DESCRIPTION
**Pull request by beyackle**
## Description

This adds awareness of the project and skill structure to the LG, LU, and QnA pages. They will now display the dialogs of whichever the current selected bot or skill is, keeping that around in the URL and using it to correctly point to the right place back in Design View.

## Task Item

closes #4615

### CODE REVIEW

File: Composer/packages/client/src/pages/language-generation/LGPage.tsx [hunk 1]
```diff
@@ -22,19 +22,25 @@ const CodeEditor = React.lazy(() => import('./code-editor'));
...
+  skillId?: string;
 }
 
 const LGPage: React.FC<RouteComponentProps<LGPageProps>> = (props: RouteComponentProps<LGPageProps>) => {
-  const { dialogId = '', projectId = '' } = props;
-  const dialogs = useRecoilValue(validateDialogsSelectorFamily(projectId));
+  const { dialogId = '', projectId = '', skillId = null } = props;
+  const dialogs = useRecoilValue(validateDialogsSelectorFamily(skillId ?? projectId));
+
+  console.log(skillId, projectId, skillId ?? projectId, dialogs);
```
Comment (srinaath) [inline Composer/packages/client/src/pages/language-generation/LGPage.tsx:32]: remove logs

File: Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx [hunk 2]
```diff
@@ -25,14 +25,18 @@ const CodeEditor = React.lazy(() => import('./code-editor'));
...
   dialogId?: string;
+  skillId?: string;
 }
 
 const QnAPage: React.FC<QnAPageProps> = (props) => {
-  const { dialogId = '', projectId = '' } = props;
+  const { dialogId = '', projectId = '', skillId = null } = props;
+
+  const baseURL =
+    projectId === skillId || skillId == null ? `/bot/${projectId}/` : `/bot/${projectId}/skill/${skillId}/`;
```
Comment (srinaath) [inline Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx:35]: projectId === skillId should never happen correct?

**Review by srinaath — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/pages/language-generation/LGPage.tsx [hunk 3]
```diff
@@ -22,19 +22,25 @@ const CodeEditor = React.lazy(() => import('./code-editor'));
...
 
   const path = props.location?.pathname ?? '';
 
   const edit = /\/edit(\/)?$/.test(path);
 
+  const baseURL =
+    projectId === skillId || skillId == null ? `/bot/${projectId}/` : `/bot/${projectId}/skill/${skillId}/`;
```
Comment (srinaath) [inline Composer/packages/client/src/pages/language-generation/LGPage.tsx:39]: same issue here projectId===skillId is a scenario that should never occur

File: Composer/packages/client/src/pages/language-understanding/LUPage.tsx [hunk 4]
```diff
@@ -21,17 +21,20 @@ const CodeEditor = React.lazy(() => import('./code-editor'));
 const LUPage: React.FC<RouteComponentProps<{
   dialogId?: string;
   projectId: string;
+  skillId: string;
 }>> = (props) => {
-  const { dialogId = '', projectId = '' } = props;
-  const dialogs = useRecoilValue(validateDialogsSelectorFamily(projectId));
+  const { dialogId = '', projectId = '', skillId = null } = props;
```
Comment (srinaath) [inline Composer/packages/client/src/pages/language-understanding/LUPage.tsx:26]: is there a reason why skillId is instanitated to null and other props are'nt? Can we make the instantiation consistent?

Comment (beyackle) [reply Composer/packages/client/src/pages/language-understanding/LUPage.tsx:25]: I originally had this as `''`, but that broke the `??` melding I was doing. If anything, I think they should all default to `null`, so I'll make that change too.

File: Composer/packages/client/src/pages/language-generation/LGPage.tsx (hunk 1 above)
Comment (beyackle) [reply Composer/packages/client/src/pages/language-generation/LGPage.tsx:31]: Oops. Thanks for catching this.

File: Composer/packages/client/src/pages/language-understanding/LUPage.tsx (hunk 4 above)
Comment (beyackle) [reply Composer/packages/client/src/pages/language-understanding/LUPage.tsx:25]: Update on this - changing them all to `null` caused more typechecking issues downstream, so I found another solution that looks less strange than this. We need a projectId and dialogId in all cases, but the skillId is optional and gets marked as such.

File: Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx [hunk 5]
```diff
@@ -21,17 +21,18 @@ import TableView from './table-view';
 
 const CodeEditor = React.lazy(() => import('./code-editor'));
 
-interface QnAPageProps extends RouteComponentProps<{}> {
-  projectId?: string;
+const QnAPage: React.FC<RouteComponentProps<{
```
Comment (hatpick) [inline Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx:24]: RouteComponentProps<TParams> makes the passed TParams partial, so all of them will be optional.

**Review by hatpick — COMMENTED** [review]
Please use:
```
`${baseUrl}restOfUrl`
```
instead of ```+``` concat.

**Review by srinaath — APPROVED** [review]

**Pull request merged**